### ISSUE TITLE
learning rate scheduler for adamp?

### ISSUE THREAD

**Issue by twmht** [clovaai/AdamP#15]
Hi,

From the paper you use cosine learning rate scheduler to train imagenet, did you apply both adamp and sgdp?

what is your opinion about using constant learning rate scheduler or other  learning rate scheduler for adam series? From my experience, using step decay or cosine would be better than constant learning rate scheduler. 

However, the advantage of adaptive optimizer is that we don't need to manually tune the learning rate scheduler,  if we need to tune the learning rate scheduler, then what is the advantage for adam series?

Here are good [discussions](https://stackoverflow.com/questions/39517431/should-we-do-learning-rate-decay-for-adam-optimizer)

thank you

**Comment by bhheo** [comment]
Hi, twmht

Thank you for sharing the interesting aspect.
I was surprised because one of my friends asks me the same question a few weeks ago.

I think Adam optimizer can not replace the learning rate scheduler since it is not designed for it.
What Adam do is measure the magnitude of gradients and divide gradients by magnitude for normalization.
Adam utilizes exponential moving average and RMS for it. it is very similar to RMSprop.
It doesn't adaptively reduce optimization step-size like lr scheduler, instead, it tries to make constant step-size invariants to the gradient scale.

The advantage of Adam is an invariants step-size to gradients scale.
For example, consider following two losses
`loss1 = 10 * f(x)`  `loss2 = 0.01 * f(x)`
If you are using SGD, `lr=0.1` with `loss1` is corresponds to `lr=100` with `loss2` 
because gradients are proportional to loss and step-size is proportional to gradients.
In other words, if the loss scale is changed then you have to tune the learning rates.
If you are using Adam, you don't need to do it.
`lr=0.001` with `loss1` is almost the same as `lr=0.001` with `loss2`.
So, with Adam optimizer, you can use the same learning rates for other scale loss functions.
In a real scenario, Adam is not 100% free from lr tuning, but the tuning scales of lr are much smaller than SGD.

An additional advantage is that this adaptive scale is applied to each parameter separately.
So, when gradients for a parameter are too small or too big, Adam automatically adjusts the gradients to make a constant step-size.
As a result, Adam mitigates gradients explosion or vanishing problems in some parameters.

Although, the auto-rate tuning phenomenon (https://arxiv.org/abs/1812.03981) reduces the learning rates like lr scheduler
I think it is not enough to replace the manual lr schedulers in general scenarios.
So, we propose AdamP to reduce this phenomenon and strictly follow the manual lr schedule.
But, I think some generative tasks use this phenomenon as a native lr scheduler so, they can work well with a constant lr scheduler.

In short, adaptive in Adam is not lr schedule so, it can't replace lr scheduler.
I know many people are confused but, the design principle is completely different from lr scheduler.
So, for the best performance, we still have to use lr scheduler even with Adam optimizer.

**Comment by twmht** [comment]
this is very helpful. thank you.

Another question is that how you select the learning rate scheduler? I tried to train resnet18 using AdamP on CIFAR100 with step decay scheduler, which is the same as SGD. but I got much more worse result than SGDM

I also used lr range test as described in https://arxiv.org/abs/1708.07120, and use OneCycle, but the result is still worse than SGDM. 

I have tried weight decay for both 1e-1 and 1e-2, but the differences of the accuracies are minor.

The best accuracies I got for SGDM is 77.6, but for AdamP is 76.4.

Can you share your experience for how to select these hyper parameters?

**Comment by bhheo** [comment]
@twmht 
Sorry for the late response. 
I thought the issue was closed because of the email alert.

I recommend a cosine learning rate scheduler without cycles.
https://github.com/rwightman/pytorch-image-models/blob/master/timm/scheduler/cosine_lr.py
As I know, many training recipes use cosine lr as their default setting and it is a good choice for image classification.
Sometimes, simple linear lr (`lr = args.lr * (1 - epoch / max_epoch)`) works better than cosine lr.
OneCycle is an innovative approach, but, I think it is not common for training recipes such as ImageNet.

However, even with cosine lr, it is hard to achieve higher performance than SGD with Adam variants.
I conducted a large number of experiments in CIFAR, but, I think SGD is much easier to use for CIFAR training.

Anyway, here is my recommendation for AdamW variants in CIFAR dataset.
For CIFAR, SGD usually uses `lr=0.1 / wd =5e-4`.
For AdamW series,  `lr=0.001 / wd=0.5` is a fair setting in an aspect of weight decay.

The weight decay is applied as multiplication with lr in common PyTorch implementation.
In SGD, weights decay with `lr * wd = 5e-5` ratio in every iteration.
But, SGD uses momentum `0.9` update without `x (1-momentum)` term. 
So, weights decay with `lr * wd / (1-momentum) = 5e-4` ratio in every iteration.
In AdamW variants, weights decay with `lr * wd` ratio.
So, with the famous `lr=0.001`, `wd` should be `0.5`. 

I hope it is not too late to help your research

Best
Byeongho Heo
